contracts/libs/Ibc.sol

```diff
@@ -172,17 +172,29 @@ library IbcUtils {
 
     // convert params to UniversalPacketBytes with optimal gas cost
 
-    function toUniversalPacketBytes(UniversalPacket memory data) internal pure returns (bytes memory packetBytes) {
-        packetBytes = abi.encode(data);
-    }
-
     // fromUniversalPacketBytes converts UniversalPacketDataBytes to UniversalPacketData, per how its packed into bytes
-    function fromUniversalPacketBytes(bytes memory data)
-        internal
+    function fromUniversalPacketBytes(bytes calldata data)
+        public
```

Comment:
Any reason from internal to public?

Reply:
The `data` argument needs to be passed in as calldata to let us use `data[96:data.length]` to set the message string dynamically - solidity doesn't let you access arrays like this for non-calldata bytes. The alternative to accessing the indicies of the array would be to write a for loop in assembly which would probably be a lot more complex and not as gas optimized (also there are [vulns](https://medium.com/@JohnnyTime/solidity-smart-contract-unbounded-loops-dos-attack-vulnerability-explained-with-real-example-f4b4aca27c08) of writing for loops of unbounded length). 

Given the above implies we have to pass in `data` as calldata now, we have to make `fromUniversalPacketBytes` an external function; since solidity only passes in calldata for external functions (otherwise solidity would pass in the data as memory and that would throw a compiler error). 

Actually, I think we need to mark this as external rather than public; since I don't think there's a way to use this as an internal function from inheriting this library